### PR TITLE
Fix for localized windows editions in testcase fn read_link() Issue#93211

### PR DESCRIPTION
**Pull request by m1guelperez**
This PR aims to fix the issue with localized windows versions that do not necessarily have the folder "Documents and settings" in English.

The idea was provided by @the8472. We check if the "CI" environment variable is set, then we always check for the "Documents and Settings"-folder, otherwise we check if the folder exists on the local machine, and if not we skip this assert.

Resoles #93211.

### CODE REVIEW

File: library/std/src/fs/tests.rs [hunk 1]
```diff
@@ -192,6 +193,7 @@ fn file_test_io_seek_and_tell_smoke_test() {
         tell_pos_post_read = check!(read_stream.seek(SeekFrom::Current(0)));
     }
     check!(fs::remove_file(filename));
+
```
Comment (hkratz) [inline library/std/src/fs/tests.rs:196]: This seems to be unrelated. Can you remove the added lines?

File: library/std/src/fs/tests.rs [hunk 2]
```diff
@@ -901,12 +907,26 @@ fn symlink_noexist() {
...
+            ),
+            Err(_e) => {
+                if Path::new(r"C:\Documents and settings\").exists() {
+                    assert_eq!(
+                        check!(fs::read_link(r"C:\Documents and settings\")),
+                        Path::new(r"C:\Users")
+                    )
+                }
+            }
+        }
```
Comment (hkratz) [inline library/std/src/fs/tests.rs:929]: Probably simpler to do something like this: 
```rust
if Path::new(r"C:\Documents and settings\").exists() || env::var_os("CI").is_some() {
    assert_eq!(
        check!(fs::read_link(r"C:\Documents and settings\")), Path::new(r"C:\Users"));
}
```
Also some comment explaining that C:\Documents and settings\ sometimes does not exist on localized Windows would be good IMHO.

**Review by hkratz — COMMENTED** [review]
Thanks for working on this!

You need to make sure that your PR does not contain any merge commits. Rebasing your branch on master should automatically remove the merge commit. See the [Git documentation](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) for details.

P.S. This is not an "official" review, I am just a contributor, not a team member.

**Comment by m1guelperez** [comment]
@hkratz Thanks for the first tips.
Do I have to create a separate PR when I rebased or is it possible to update this one?

**Comment by the8472** [comment]
A force-push will update the PR. See also the relevant dev guide sections:
https://rustc-dev-guide.rust-lang.org/git.html#no-merge-policy
https://rustc-dev-guide.rust-lang.org/git.html#rebasing

**Comment by m1guelperez** [comment]
Sorry I am stuck. 

I can not pull changes from the rust master (not my fork).
While executing: ```git pull upstream master --ff-only``` I get the following error:

```
From https://github.com/rust-lang/rust
 * branch                master     -> FETCH_HEAD
fatal: Not possible to fast-forward, aborting.
```

**What I did before:**
On my fork, I created an issue branch, with:
```git checkout -b issue-93211-fix```

 where I updated the fix and pushed it back to my master with:

```git add .\library\std\src\fs\tests.rs```
```git commit -m "Resolved typos and deleted unused variable"```
```git push```
```git checkout master```
```git rebase issue-93211-fix```

Which ended with:
```Successfully rebased and updated refs/heads/master.```

In MY git repo from the fork I now find the following:

```This branch is 4 commits ahead, 40 commits behind rust-lang:master.```
So before I commit to my master branch now, I should rebase as far as I understood.

However if I now execute:
```git rebase master``` I will get: ```Current branch master is up to date.```
If I try to pull the changes from the rust-lang:master with:
```git pull upstream master --ff-only```

I get:
```remote: Enumerating objects: 15, done.
remote: Counting objects: 100% (10/10), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 15 (delta 7), reused 8 (delta 7), pack-reused 5
Unpacking objects: 100% (15/15), 18.45 KiB | 331.00 KiB/s, done.
From https://github.com/rust-lang/rust
 * branch                    master     -> FETCH_HEAD
   df368ae457c..17dfae79bbc  master     -> upstream/master
fatal: Not possible to fast-forward, aborting.
```

I am not sure how to resolve this, without making my branch just more of a mess.
If it would be easier for you, you can also hit me up on Zulip, username: Anon

**Comment by the8472** [comment]
My rebase workflow looks like this:

```bash
git switch issue-93211-fix # change to branch, if already on it
git fetch upstream # get the most recent commits from the rust-lang repo rather than your own github fork
git rebase upstream/master # rebase on the remote master branch, not your local master, you can update the local master later
# fix conflicts here if necessary
git push --force # overwrite remote branch on your github fork with your local branch
```

`git pull` is only needed to get your local master branch in sync with the remote one, it also does a fetch.

To visualize current git state I have this in my  `.gitconfig`, which can then be invoked as `git graph`:

```
[alias]
graph = log --graph --all --oneline --date=format:'%Y-%m-%d %H:%M' --format=format:'%C(auto)%h %C(green)%cd%C(auto) %s%d'
```

**Comment by m1guelperez** [comment]
> My rebase workflow looks like this:
> 
> ```shell
> git switch issue-93211-fix # change to branch, if already on it
> git fetch upstream # get the most recent commits from the rust-lang repo rather than your own github fork, without updating any local branches
> git rebase upstream/master # rebase on the remote master branch, not your local master, you can update the local master later
> # fix conflicts here if necessary
> git push --force # overwrite remote branch on your github fork with your local branch
> ```

Okay, I followed your steps till here and everything went fine so far.

Now I have to fix ```This branch is 4 commits ahead, 40 commits behind rust-lang:master.``` on my master branch somehow correct?

File: library/std/src/fs/tests.rs (hunk 1 above)
Comment (m1guelperez) [reply library/std/src/fs/tests.rs:196]: I deleted the unused line.

File: library/std/src/fs/tests.rs [hunk 3]
```diff
@@ -204,15 +205,18 @@ fn file_test_io_seek_and_write() {
     let overwrite_msg = "-the-bar!!";
     let final_msg = "foo-the-bar!!";
     let seek_idx = 3;
+
```
Comment (hkratz) [inline library/std/src/fs/tests.rs:208]:
```suggestion
```

File: library/std/src/fs/tests.rs [hunk 4]
```diff
@@ -204,15 +205,18 @@ fn file_test_io_seek_and_write() {
...
     let mut read_mem = [0; 13];
     let tmpdir = tmpdir();
     let filename = &tmpdir.join("file_rt_io_file_test_seek_and_write.txt");
     {
         let mut rw_stream = check!(File::create(filename));
+
```
Comment (hkratz) [inline library/std/src/fs/tests.rs:214]:
```suggestion
```

File: library/std/src/fs/tests.rs [hunk 5]
```diff
@@ -204,15 +205,18 @@ fn file_test_io_seek_and_write() {
...
         check!(rw_stream.write(initial_msg.as_bytes()));
         check!(rw_stream.seek(SeekFrom::Start(seek_idx)));
         check!(rw_stream.write(overwrite_msg.as_bytes()));
     }
+
```
Comment (hkratz) [inline library/std/src/fs/tests.rs:219]:
```suggestion
```

File: library/std/src/fs/tests.rs [hunk 6]
```diff
@@ -226,6 +230,7 @@ fn file_test_io_seek_and_write() {
 fn file_test_io_seek_shakedown() {
     //                   01234567890123
     let initial_msg = "qwer-asdf-zxcv";
+
```
Comment (hkratz) [inline library/std/src/fs/tests.rs:233]:
```suggestion
```

**Review by hkratz — COMMENTED** [review]
Those lines are also not part of your change.

**Comment by m1guelperez** [comment]
@hkratz I think it should be correct now, hopefully, it went well this time.

**Comment by Mark-Simulacrum** [comment]
It seems unfortunate to have a CI gate on this test; I wonder if there's a better path that is not localized (for example) that we could use.

@rustbot ping windows -- is there a better path for us to use here? Is there something we could check (better than a CI variable) to confirm that we're on an English-localized system perhaps?

**Comment by m1guelperez** [comment]
> It seems unfortunate to have a CI gate on this test; I wonder if there's a better path that is not localized (for example) that we could use.
> 

Thanks for the feedback so far! 
I feel like there are several workarounds, but none is "really nice". 

First, there is the one with the CI gate, which I actually liked.

But with admin rights, we could also create the folder and instantly delete it after the test is over.
However, this needs admin rights iirc. 

An absolute overkill solution would be to make a map, which contains the "Documents and Settings" folder in every language and we then simply make a lookup.

I googled a bit to see if we could easily look in which language the initial windows installation was, but I did not find anything valuable.

But yes, your idea regarding a non-localized path would be perfect!

File: library/std/src/fs/tests.rs [hunk 7]
```diff
@@ -906,7 +907,14 @@ fn read_link() {
         // junction
         assert_eq!(check!(fs::read_link(r"C:\Users\Default User")), Path::new(r"C:\Users\Default"));
         // junction with special permissions
-        assert_eq!(check!(fs::read_link(r"C:\Documents and Settings\")), Path::new(r"C:\Users"));
+        // Since not all localized windows versions contain the folder "Documents and settings" in english,
+        // we will briefly check, if it exists and otherwise skip the test. Except during CI we will always execute the test.
+        if Path::new(r"C:\Documents and settings\").exists() || env::var_os("CI").is_some() {
```
Comment (Mark-Simulacrum) [inline library/std/src/fs/tests.rs:912]: It looks like this might've lowercased Settings by accident? It should probably be "Settings" still. (Also in the assert).

**Comment by Mark-Simulacrum** [comment]
I'm inclined to move ahead with this PR (almost as-is, modulo the nit), and we can reconsider if it becomes a pain point in the future -- perhaps to just dropping the test altogether.

File: library/std/src/fs/tests.rs (hunk 7 above)
Comment (m1guelperez) [reply library/std/src/fs/tests.rs:912]: Yes, this is correct. It was an accident and should be in uppercase as well.

**Comment by m1guelperez** [comment]
> I'm inclined to move ahead with this PR (almost as-is, modulo the nit), and we can reconsider if it becomes a pain point in the future -- perhaps to just dropping the test altogether.

Okay, so for now, you think it is not a big deal and therefore we shouldn't change this test currently?
Should I close this PR then or let it open?

**Comment by Mark-Simulacrum** [comment]
If you fix the nit, then I will go ahead and approve the PR; if for some reason it causes us problems we can reconsider at that point the best course of action.

**Comment by m1guelperez** [comment]
Okay just to clarify, because I am confused of the word "nit".

I guess you mean when I fix the typo regarding the "Settings" correct? 

For sure, I try to fix that asap, I have to check if my laptop is able to handle, compiling the whole compiler, since I can't access my main pc for 1 month.

I will try it out tomorrow!

**Comment by Mark-Simulacrum** [comment]
Feel free to just amend the commit; no need to test locally. If you could squash afterwards into one commit as well that would be good.

**Comment by m1guelperez** [comment]
> Feel free to just amend the commit; no need to test locally. If you could squash afterwards into one commit as well that would be good.

Hey!
I  fixed the typos and tried to squash it in one commit. Please let me know, if everything worked out, I am still not that experienced with git.

**Comment by wesleywiser** [comment]
> is there a better path for us to use here? Is there something we could check (better than a CI variable) to confirm that we're on an English-localized system perhaps?

I'm not aware of a better approach, unfortunately.

**Comment by Mark-Simulacrum** [comment]
@bors r+ rollup

Thanks for your patience @m1guelperez, this looks great. I've approved it now and it will go into the [merge queue](https://bors.rust-lang.org/queue/rust) before actually merging.

**Comment by m1guelperez** [comment]
> @bors r+ rollup
> 
> 
> 
> Thanks for your patience @m1guelperez, this looks great. I've approved it now and it will go into the [merge queue](https://bors.rust-lang.org/queue/rust) before actually merging.

No problem! Glad to help and can't wait to make my next contribution!

Thanks to the rust community for being always helpful especially as a beginner! 

See you. :-)
